### PR TITLE
Improvements to IOSim compatibility

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This PR significantly simplifies the IOSim compatibility module by re-organizing `StateModel` to differentiate between the "model type" and the "perform type". One big plus of this is that we've won the ability to have models with different backends where the types might even by symbolic! (One can always hope!)

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
@abailly Have a look at this - I think it's pretty cool how this works out!

**Comment by MaximilianAlgehed** [comment]
And yes, this is a major breaking change and we need to discuss how to best handle that. I also haven't finished everthing, but I think it's good getting a bit of a spotlight on this.

File: quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs
Comment (abailly-iohk) [inline quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:132]: Question: What's the rationale behind introducing this type family?

File: quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs
Comment (abailly-iohk) [inline quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:152]: Question: Why move these from `StateModel`? It seems to me the `postcondition` on a state transition should be a property of the _model_ not of its execution. This is less clear for `monitoring` and I agree it might make sense to make it part of the execution.

**Comment by MaximilianAlgehed** [comment]
> Could you elaborate on the reason behind those changes? I am a bit confused because we already use the `StateModel` in hydra along with `IOSim` and it works fine. Perhaps we are doing things in a naive way?

Have a look at the changes to the registry example. To be able to write properties where the state isn't indexed by s (which you need to be able to work properly with DL properties) you need the state type not to mention the s parameter.

File: quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs
Comment (MaximilianAlgehed) [reply quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:189]: This is just a leftover from experimenting. Should be removed.

File: quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs
Comment (MaximilianAlgehed) [reply quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:132]: C.f. discussion below about wanting different backends with different concrete representations of the things referenced in the model. Important if you want to use a model state that refers to things that would be indexed by the `s` in the state model type otherwise.

See the Registry example for a showcase of why this matters.

Comment (MaximilianAlgehed) [reply quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:132]: Note that an example of this is `action` in the `DL` monad - it requires the result of a state model action to be `Typeable` - but if you want the result to be a `ThreadId (IOSim s)` you can't do it because `s` can't be typeable if you want to actually run your property.

File: quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs
Comment (abailly-iohk) [reply quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:152]: Ok, so the reason why we (in Hydra) did not encounter this issue is because we don't have any result type depending on the actual monad the model is executed in. Is that correct? And you are having this issue because you need to have data in the Registry model that depends on `s`, namely `ThreadId`?

Comment (abailly-iohk) [reply quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:152]: So my next question would be: Isn't there another way to model the `Registry` that would not require introducing this indirection?

Comment (MaximilianAlgehed) [reply quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs:152]: You're getting it yes.

You could model it differently, but that would be by manually introducing the indirection. But since we already have the Var indirection we piggy back on that.

We had the old IOSim compat module that did this but it was ugly and unnecessary.

**Comment by MaximilianAlgehed** [comment]
@abailly-iohk do you know why the build isn't working?

**Comment by abailly-iohk** [comment]
It seems zlib-dev is missing but I don't understand why it fails now and wasn't before, you did not change anything in the dependencies. Adding `mtl` should not entail recompilation of the whole world. Don't bother with this, I can take care of turning it green later, once you're happy with the code and it builds locally.

**Comment by MaximilianAlgehed** [comment]
I think I'm pretty happy with the code now. I'll just bump the version and we can merge this.

File: quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs
Comment (abailly-iohk) [inline quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs:53]: nitpick: We should settle on a single code formatter in order to avoid these kind of changes. The default provided by the nix-shell is fourmolu with some configuration. I am fine with any other formatter so long as there's only one. We should probably put a "coding standards" somewhere too.

File: quickcheck-dynamic/quickcheck-dynamic.cabal
Comment (abailly-iohk) [inline quickcheck-dynamic/quickcheck-dynamic.cabal:3]: Could you also update the `CHANGELOG.md` file?

**Review by abailly-iohk — CHANGES_REQUESTED** [review]
I am still not entirely convinced we need the `Realized` typeclass to be part of the library: The `Model` should be completely independent of whatever `Monad` instance the `RunModel` is expressed in, and it should the latter's role to provide a translation layer which is specific to _this_ execution engine. But I feel you know better so I will defer to your expertise.
Could you please update the CHANGELOG.md with some high-level statement about this change and also format the code with `fourmolu` for consistency's sake?

File: quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs
Comment (MaximilianAlgehed) [reply quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs:53]: Petition to make a separate PR right after this one that fixes formatting and introduces a `formolu` standard? Perhaps we should also put it in CI to be sure?

File: quickcheck-dynamic/quickcheck-dynamic.cabal
Comment (MaximilianAlgehed) [reply quickcheck-dynamic/quickcheck-dynamic.cabal:3]: Will do.

**Comment by MaximilianAlgehed** [comment]
@abailly-iohk I understand the concern about the `Realized` type family and splitting the postcondition out. However, I think the main arugments for this change are:
1. There is already a level of indirection in the library (`Var`) and you don't want to force every new `perform` to introduce its own indirection tricks to deal with this situation.
2. You want to make it possible to have many different runners - e.g. `ST`, `IO`, and `IOSim` or a concolic / symbolic testing engine and splitting out the postcondition allows you to have something like a *symbolic* postcondition that does something like SMT magic or an `IO` postcondition that queries the actual state of the system.

**Comment by MaximilianAlgehed** [comment]
Also the model *is* now completely independent of the monad that `RunModel` is expressed in!

What we _could_ add back is a *model* postcondition. But since this would have to talk about the *ideal* return type of an action it's not clear what this would mean. It's certainly not clear that you can always obtain a value of the ideal return type (e.g. in a symbolic or `IOSim` runner).

File: quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs
Comment (MaximilianAlgehed) [reply quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs:53]: I ran `fourmolu` in `nix-shell` (without any input - how works?!) and nothing happened... I then ran:
```
find . -type f -name *.hs | xargs -n1 fourmolu
```
and it just spat out a bunch of stuff on `stdout`. Is there some unknown magic I need to run to make it actually format every file in place?

**Comment by abailly-iohk** [comment]
Interestingly, we now have a failing test: https://github.com/input-output-hk/quickcheck-dynamic/runs/8183759215?check_suite_focus=true#step:7:36 🎉 🤔

File: quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs
Comment (abailly-iohk) [reply quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs:53]: Yeah, it needs an additional argument. I agree we should leave it as it is now and add something more automated later :)

**Comment by MaximilianAlgehed** [comment]
> Interestingly, we now have a failing test: https://github.com/input-output-hk/quickcheck-dynamic/runs/8183759215?check_suite_focus=true#step:7:36 🎉 🤔

Yes! That's on purpose ;) We could disable that test for now and I'll submit a new PR that turns it on. A fix to this failing test will hopefully be merged in `io-sim` soon:
https://github.com/input-output-hk/io-sim/pull/19

File: quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs
Comment (MaximilianAlgehed) [reply quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs:53]: Do you know what the additional argument is? ;)

**Comment by abailly-iohk** [comment]
Understood. But then we can expect the failure so that it now passes and when iosim gets updated, it will fail again?

**Comment by MaximilianAlgehed** [comment]
When `io-sim` is updated we need to change the implementation of the `Registry` example a bit because `io-sim` is currently lacking `threadStatus` and we need it to correctly implement the example. So nothing will be "automatic" as it were. Easiest solution here is just commenting it out and as soon as the other PR is merged in `io-sim` I make the necessary changes to the example and add in the tests.

File: quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs
Comment (abailly-iohk) [reply quickcheck-dynamic-iosim/test/Spec/DynamicLogic/RegistryModel.hs:53]: `-i` (like _in-place_, the long version is `--mode inplace`)

**Review by abailly-iohk — APPROVED** [review]

**Comment by MaximilianAlgehed** [comment]
Waiting for it to turn green and then I'll merge and get started on the `fourmolu` PR.

**Pull request merged**
